File: klaverjas/game/cards.py
from enum import Enum, unique


@unique
class Suit(Enum):
    SPADES = 0
    HEARTS = 1
    CLUBS = 2
    DIAMONDS = 3

    def __str__(self):
        return ['spades', 'hearts', 'clubs', 'diamonds'][self.value]


@unique
class Rank(Enum):
    SEVEN = 0
    EIGHT = 1
    NINE = 2
    TEN = 3
    JACK = 4
    QUEEN = 5
    KING = 6
    ACE = 7

    def __str__(self):
        return ['7', '8', '9', '10', 'J', 'Q', 'K', 'A'][self.value]


class Card:
    def __init__(self, suit, rank):
        if not isinstance(suit, Suit):
            raise TypeError
        if not isinstance(rank, Rank):
            raise TypeError
        self._suit = suit
        self._rank = rank

    def __eq__(self, other):
        return self._suit == other._suit and self._rank == other._rank

    def __hash__(self):
        return hash(repr(self))

    def __repr__(self):
        return '{}{}'.format(self._suit, self._rank)

    def order(self, trump_suit):
        if self._suit == trump_suit:
            return [8, 9, 14, 12, 15, 10, 11, 13][self._rank.value]
        return [0, 1, 2, 6, 3, 4, 5, 7][self._rank.value]

    def points(self, trump_suit):
        if self._suit == trump_suit:
            return [0, 0, 14, 10, 20, 3, 4, 11][self._rank.value]
        return [0, 0, 0, 10, 2, 3, 4, 11][self._rank.value]

    def rank(self):
        return self._rank

    def suit(self):
        return self._suit

    def to_dict(self):
        return {'rank': str(self._rank), 'suit': str(self._suit)}


def card_from_dict(dict):
    return Card(suit_from_label(dict['suit']), rank_from_label(dict['rank']))


def suit_from_label(label):
    return Suit(['spades', 'hearts', 'clubs', 'diamonds'].index(label))


def rank_from_label(label):
    return Rank(['7', '8', '9', '10', 'J', 'Q', 'K', 'A'].index(label))

File: klaverjas/static/scripts/cards.js
import {Animations} from "./animations.js";


const RANKS = {
    SEVEN: {
        html: "7",
        label: "seven",
        value: 0
    },
    EIGHT: {
        html: "8",
        label: "eight",
        value: 1
    },
    NINE: {
        html: "9",
        label: "nine",
        value: 2
    },
    TEN: {
        html: "10",
        label: "ten",
        value: 3
    },
    JACK: {
        html: "J",
        label: "jack",
        value: 4
    },
    QUEEN: {
        html: "Q",
        label: "queen",
        value: 5
    },
    KING: {
        html: "K",
        label: "king",
        value: 6
    },
    ACE: {
        html: "A",
        label: "ace",
        value: 7
    },
};

const SUITS = {
    SPADES: {
        html: "&spades;&#xfe0e;",
        label: "spades",
        value: 0
    },
    HEARTS: {
        html: "&hearts;&#xfe0e;",
        label: "hearts",
        value: 1
    },
    CLUBS: {
        html: "&clubs;&#xfe0e",
        label: "clubs",
        value: 2
    },
    DIAMONDS: {
        html: "&diams&#xfe0e",
        label: "diamonds",
        value: 3
    }
};


class Card {
    constructor($parent, hidden, suit, rank) {
        this.flipped = hidden;
        this.rank = rank;
        this.suit = suit;
        this.x = 0;
        this.y = 0;
        this.angle = 0;
        this.$el = document.createElement("div");

        this.$el.className = "card";
        const $back = document.createElement("div");
        $back.className = "back";
        if (!hidden) {
            const $face = document.createElement("div");
            const $rank = document.createElement("span");
            const $suit = document.createElement("span");

            $face.className = "face suit-" + suit.label;
            $rank.className = "rank";
            $suit.className = "suit";

            $rank.innerHTML = rank.html;
            $suit.innerHTML = suit.html;

            $face.appendChild($rank);
            $face.appendChild($suit);
            this.$el.appendChild($face);
        } else {
            this.$el.style.transform = "rotateY(180deg)";
        }
        this.$el.appendChild($back);
        //this.$el.style.visibility = "hidden";
        $parent.appendChild(this.$el);
    }

    animate_to(delay, duration, x, y, angle) {
        const start_x = this.x;
        const start_y = this.y;
        const start_angle = this.angle;
        const d_x = x - this.x;
        const d_y = y - this.y;
        const d_angle = angle - this.angle;
        const card = this;

        Animations.create(duration,
            function(d_t) {
                card.x = d_x * d_t + start_x;
                card.y = d_y * d_t + start_y;
                card.angle = d_angle * d_t + start_angle;
                card.$el.style.transform = "translate(" + card.x + "px, " + card.y + "px) rotate(" + card.angle + "deg)" + (card.flipped ? " rotateY(180deg)" : "");
            }, {effect: Animations.effects.ease, delay: delay});
    }

    animate_flip(delay, duration) {
        const card = this;
        Animations.create(duration,
            function(d_t) {
                const angle = card.flipped ? 180 - 180 * d_t : 180 * d_t;
                card.$el.style.transform = "translate(" + card.x + "px, " + card.y + "px) rotate(" + card.angle + "deg)" + (card.flipped ? " rotateY(" + rot + "deg)" : "");
            }, {effect: Animations.effects.ease, on_end: function() { card.flipped = !card.flipped; }, delay: delay});
    }
}


export {
    Card,
    RANKS,
    SUITS
}
